### PR TITLE
Shift site build to use monorepo and gh-pages

### PR DESCRIPTION
**Pull request by bitsondatadev**
Revert back to using the `site` directory which contains the monorepo of the site and versioned docs. Adds in the site styling to look like the current Apache Iceberg site. Adds the main build infrastructure.

I gleaned info from [this PR](https://github.com/apache/iceberg/pull/4062/files), [this PR](https://github.com/apache/iceberg/pull/4045), [this PR](https://github.com/apache/iceberg-docs/pull/36/files) and a bit off of [this deploy script](https://github.com/apache/iceberg-docs/blob/main/.github/workflows/deploy.yml).

Here is a link to the gh-pages in my branch: https://bitsondatadev.github.io/iceberg/

I will be doing another PR to handle the following task items. when we get some eyes on the current site and if anyone sees any issues.

Also move back to using the asf infra standards including the [asf-site](https://github.com/apache/iceberg/tree/asf-site) branch.

This PR will be moving `/docs/docs/nightly/` up to `/docs/nightly/` based on some bad state that happened when we pulled the versioned docs into the worktree. So most of the changed files are just those files being moved.

### CODE REVIEW

File: .github/workflows/spark-ci.yml [hunk 1]
```diff
@@ -141,4 +142,4 @@ jobs:
         with:
           name: test logs
           path: |
-            **/build/testlogs
\ No newline at end of file
+            **/build/testlogs
```
Comment (nastra) [inline .github/workflows/spark-ci.yml:145]: nit: unnecessary change

Comment (bitsondatadev) [reply .github/workflows/spark-ci.yml:145]: So I just tried removing the file changes from my local and the GH PR, then added this line again as I did in the other files and this artifact keeps popping up in the diff. Any thoughts on how to get rid of it? It's not a newline or even a space or anything. I tried those as well.

Comment (Fokko) [reply .github/workflows/spark-ci.yml:145]: Looks like the trailing newline was missing.

File: .github/labeler.yml [hunk 2]
```diff
@@ -32,6 +32,7 @@ BUILD:
   - "**/*gradle*"
 DOCS:
   - docs/**/*
+  - site/**/*
```
Comment (rdblue) [inline .github/labeler.yml:35]: Can we fix this in a separate PR? It isn't necessary for this one and it would be great to get it in faster.

File: site/README.md [hunk 3]
```diff
@@ -65,7 +65,7 @@ The non-versioned site pages are all the `/site/docs/.*md` files and the docs ar
 > This build process is currently missing older versions and the javadoc branches.
 > Until these branches are merged, these steps will not work.
 
-All previously versioned docs will be committed in `docs-<version>` branches and mounted using [git worktree](https://git-scm.com/docs/git-worktree) at build time. The worktree will pull these versions in following the `/site/docs/docs/<version>` convention. The `latest` version, will be a secondary copy of the most recent build version in the worktree, but pointing to `/site/docs/docs/latest`. There is also a `javadoc` branch that contains all prior static generation versions of the javadocs in a single tag.
+The Iceberg versioned docs are committed in the `docs` branch and mounted using [git worktree](https://git-scm.com/docs/git-worktree) at build time. The worktree pulls these version directories in following the `/site/docs/docs/<version>` convention. The `latest` version, will be a secondary copy of the most recent build version in the worktree, but pointing to `/site/docs/docs/latest`. There is also a `javadoc` branch that contains all prior static generation versions of the javadocs.
```
Comment (rdblue) [inline site/README.md:68]: Are all docs in the same branch?

Comment (bitsondatadev) [reply site/README.md:68]: Yes, I moved to this assumption so we don't have to make branches per version anymore. I think that will make changes across multiple docs much faster.

File: site/README.md [hunk 4]
```diff
@@ -125,17 +124,16 @@ mkdocs serve
 ## Release process
 
 Deploying a version of the docs is a two step process:
- 1. ~~Cut a new release from the current branch revision. This creates a new branch `docs-<version>`.~~
+ 1. ~~Release a new version by copying the current nightly directory to a new version directory in the `docs` branch.~~
 
     ```
     .github/bin/deploy_docs.sh -v 1.4.0
```
[the remark this answers is not in the document]
Comment (bitsondatadev) [reply site/README.md:130]: Yeah,
I had a discussion with @rdblue that we should add deploy scripts in this PR and not have the script that I added where if run in the wrong directory did bad things 😬.

https://github.com/apache/iceberg/pull/8659#discussion_r1367561171

File: site/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (bitsondatadev) [reply site/README.md:68]: Not that this isn't correct English but as a native speaker, it reads almost like you're saying the convention is what pulls the directories in.

File: site/README.md [hunk 5]
```diff
@@ -125,17 +124,16 @@ mkdocs serve
 ## Release process
 
 Deploying a version of the docs is a two step process:
- 1. ~~Cut a new release from the current branch revision. This creates a new branch `docs-<version>`.~~
+ 1. ~~Release a new version by copying the current nightly directory to a new version directory in the `docs` branch.~~
```
[the remark this answers is not in the document]
Comment (bitsondatadev) [reply site/README.md:127]: https://github.com/apache/iceberg/tree/main/site/docs/docs/nightly

File: site/docs/docs/nightly/docs/configuration.md
Comment (bitsondatadev) [inline site/docs/docs/nightly/docs/configuration.md:1]: All of the changes under `site/docs/docs/**` are all in a seperate commit. This was just to keep the docs up-to-date so I could release Docs for 1.4.2.

File: site/dev/common.sh [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+uninstall_deps () {
+  if test -f .env; then
+    deactivate
+    rm -r .env
+  fi
+}
+
+install_deps () {
+  uninstall_deps
+  python -m venv .env
```
Comment (Fokko) [inline site/dev/common.sh]: This already assumes that you have `virtualenv` installed. Do we actually need a `venv` for this simple setup? We could also use the system `python3`

File: site/README.md [hunk 7]
```diff
@@ -83,59 +99,27 @@ All previously versioned docs will be committed in `docs-<version>` branches and
...
-
+Deploying the docs is a two step process:
+ 1. Release a new version by copying the current nightly directory to a new version directory in the `docs` branch.
     ```
-    .github/bin/deploy_docs.sh -v 1.4.0
+    make release ICEBERG_VERSION=${ICEBERG_VERSION}
+    ```
+ 1. Push the generated site to `asf-site`.
+    ```
+    make deploy
```
Comment (Fokko) [inline site/README.md:121]: Will this be done by a person? We could also do this in the CI, as we do with PyIceberg: https://github.com/apache/iceberg-python/blob/main/.github/workflows/python-ci-docs.yml

File: site/dev/common.sh [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+
+# https://squidfunk.github.io/mkdocs-material/setup/setting-up-site-search/#search-exclusion
+search_exclude_versioned_docs () {
+  local ICEBERG_VERSION="$1"
+  assert_not_empty "${ICEBERG_VERSION}"
+
+  cd "${ICEBERG_VERSION}/docs/"
+  
+  python3 -c "import os
```
Comment (Fokko) [inline site/dev/common.sh:88]: We both use python and python3, we probably want to consolidate this

File: site/Makefile [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+.PHONY: help
+help: # Show help for each of the Makefile recipes.
+	@grep -E '^[a-zA-Z0-9 -]+:.*#'  Makefile | sort | while read -r l; do printf "\033[1;32m$$(echo $$l | cut -f 1 -d':')\033[00m:$$(echo $$l | cut -f 2- -d'#')\n"; done
+
+.PHONY: serve
+serve: # Clean, build, and run the docs site locally.
+	dev/serve.sh
+
+.PHONY: build
+build: # Clean and build the docs site locally.
```
Comment (Fokko) [inline site/Makefile:25]: In Makefile you can also invoke other commands, for example here: https://github.com/apache/iceberg-python/blob/main/Makefile#L24

This way we can pull logic from the scripts themselves, into the Makefile (and therefore we don't need this kind of comments).

File: site/dev/common.sh [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+  if test -f .env; then
+    deactivate
+    rm -r .env
+  fi
+}
+
+install_deps () {
+  uninstall_deps
+  python -m venv .env
+  source env/bin/activate
```
Comment (Fokko) [inline site/dev/common.sh]:
```suggestion
  source .env/bin/activate
```

File: site/dev/common.sh [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+for f in filter(lambda x: x.endswith('.md'), os.listdir()): lines = open(f).readlines(); open(f, 'w').writelines(lines[:2] + ['search:\n', '  exclude: true\n'] + lines[2:]);"
+  
+  cd -
+}
+
+pull_versioned_docs () {
+  mv_nightly_up 
+  
+  git worktree add docs/docs docs
+  git worktree add docs/javadoc javadoc
```
Comment (Fokko) [inline site/dev/common.sh:98]: My `make serve` gets stuck on this one:

```
+ git worktree add docs/docs docs
Preparing worktree (checking out 'docs')
HEAD is now at 87d2dfbd6 Shift site build to use monorepo and gh-pages
+ git worktree add docs/javadoc javadoc
fatal: invalid reference: javadoc
```

File: site/.gitignore [hunk 12]
```diff
@@ -1,13 +1,3 @@
-## Temp remove for first phase
```
[the remark this answers is not in the document]
Comment (bitsondatadev) [reply site/.gitignore:1]: Yeah, fair enough...I added a few specifics from my previous one

File: site/dev/common.sh (hunk 8 above)
Comment (bitsondatadev) [reply site/dev/common.sh:88]: yeesh, yeah...I think best practice is python3 no?

File: site/Makefile (hunk 9 above)
Comment (bitsondatadev) [reply site/Makefile:25]: lol, actually, the comments were a hack for `make help` or default `make` output.

File: site/README.md (hunk 7 above)
Comment (bitsondatadev) [reply site/README.md:121]: No, I guess I should remove this as that basically documents this step. I'll only keep build and serve documented in the readme.

Although one last thought is that I just want all this documented so that others can read through to understand my intent.

File: site/dev/common.sh (hunk 8 above)
Comment (Fokko) [reply site/dev/common.sh:88]: Yes, I prefer `python3`

File: site/dev/common.sh (hunk 11 above)
Comment (bitsondatadev) [reply site/dev/common.sh:88]: Should be good now!

File: site/dev/common.sh [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
```
Comment (Fokko) [inline site/dev/common.sh:88]:
```suggestion
  git remote add iceberg-docs https://github.com/apache/iceberg.git
  git remote fetch iceberg-docs
  git worktree add docs/docs iceberg-docs/docs
  git worktree add docs/javadoc iceberg-docs/javadoc
```

**Comment by Fokko** [comment]
Niiice! Got the `make serve` running in one go now:

image

File: site/README.md [hunk 14]
```diff
@@ -35,107 +35,101 @@ In MkDocs, the [`docs_dir`](https://www.mkdocs.org/user-guide/configuration/#doc
...
+This step will generate the following layout:
 
 ```
 ./site/
 └── docs
     ├── docs
-    │   ├── nightly
-    │   ├── latest
+    │   ├── nightly (symlink to /site/docs/nightly/)
+    │   ├── latest (symlink to /site/docs/1.4.0/)
```
Comment (Fokko) [inline site/README.md:89]: Should we add the 1.4.2 etc here for clarity?

File: site/mkdocs.yml [hunk 15]
```diff
@@ -23,33 +23,37 @@ theme:
...
 nav:
   - Quickstart:
     - Spark: spark-quickstart.md
     - Hive: hive-quickstart.md
   - Docs: 
     - nightly: '!include docs/docs/nightly/mkdocs.yml'
-#    - latest: '!include docs/docs/latest/mkdocs.yml'
-#    - 1.3.1: '!include docs/docs/1.3.1/mkdocs.yml'
-#    - 1.3.0: '!include docs/docs/1.3.0/mkdocs.yml'
+    - latest: '!include docs/docs/latest/mkdocs.yml'
```
Comment (Fokko) [inline site/mkdocs.yml:53]: I was looking at this:

image

It looks a bit odd since 1.4.2 is the same as latest. Do we really need this, it is quite obvious that 1.4.2 is the latest.

File: .github/workflows/site-ci.yml [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+#
+name: site-ci 
+on:
+  push:
+    branches:
+      - main
+    paths:
+      - site/**
+permissions:
+  contents: write
```
Comment (Fokko) [inline .github/workflows/site-ci.yml:27]: Interesting, this wasn't needed for Python: https://github.com/apache/iceberg-python/blob/main/.github/workflows/python-ci-docs.yml

**Review by Fokko — APPROVED** [review]
Left some small comments, but in the greater scheme of things; LGTM

File: site/mkdocs.yml (hunk 15 above)
Comment (bitsondatadev) [reply site/mkdocs.yml:53]: I've gotten mixed signals depending who I've spoken to about meta refresh for redirects.

According to most of the stuff that comes up it's [not as big of a deal](https://webmasters.stackexchange.com/a/141240) as some made it out to be. So we can likely change `latest` to being a refresh page, but I'm curious to see how mkdocs renders that.

We should ideally have latest get ranked by Google and we may just want to be cognizant if we get a bunch of random versions or if latest gets ranked higher.

It's a bit of lore and black box when it comes to ensuring a certain version gets indexed and ranked by Google. No real way to control it so this was my feeble attempt.

<https://developers.google.com/search/docs/crawling-indexing/301-redirects#metarefresh>

Based on their docs, I think we should be fine if we update all the pages to set the [`site_url`](https://www.mkdocs.org/user-guide/configuration/#site_url) which will always point to latest. 

Also, we'll likely want to add an [outdated version warning](https://squidfunk.github.io/mkdocs-material/setup/setting-up-versioning/#version-warning) which will link to the "latest" equivalent of the page if available.

File: site/README.md (hunk 14 above)
Comment (bitsondatadev) [reply site/README.md:89]: Yeah, that or we could just use some variable placeholder representation in the README to represent the latest version: `$LATEST` or something. My only concern is people might read it too literally.

Maybe we can do the good ol' `<latest_version>` or like pseudo mustache syntax `{latest_version}`.

Any preference?

File: site/README.md [hunk 17]
```diff
@@ -35,107 +35,101 @@ In MkDocs, the [`docs_dir`](https://www.mkdocs.org/user-guide/configuration/#doc
 
 ### Iceberg docs layout
 
-In the Iceberg docs, since the top-level site and versioned docs are contained in the same directory, they all live under the `/site` directory of the main Iceberg repository. The `/site/docs` directory is named this way to follow the [MkDocs convention](https://www.mkdocs.org/user-guide/configuration/#docs_dir), while the `/site/docs/docs` directory is an analog to the "Docs" navigation tab. Under this directory, you'll find the `/site/docs/docs/nightly` directory, which contains the state of the documentation in the local revisions.
+The static Iceberg website and versioned documentation live together under the `/site` directory of the main Iceberg repository. The `/site/docs` directory is named that way to follow the [MkDocs convention](https://www.mkdocs.org/user-guide/configuration/#docs_dir), while the nested `/site/docs/docs` directory is an analog to the "Docs" navigation tab. The `/site/docs/nightly` directory contains the current state of the versioned documentation with local revisions.
```
Comment (RussellSpitzer) [inline site/README.md:38]: minor nit: "directory is named that way to follow the" => "is named following the"

File: site/README.md [hunk 18]
```diff
@@ -35,107 +35,101 @@ In MkDocs, the [`docs_dir`](https://www.mkdocs.org/user-guide/configuration/#doc
...
 
-The non-versioned site pages are all the `/site/docs/.*md` files and the docs are the `/site/docs/docs/<version>/docs/*.md` files. Notice the location of the `mkdocs.yml`. Looking at this though, you may ask where the older versions and javadocs are.
+
+The static Iceberg site pages are Markdown files that live at `/site/docs/.*md`. The versioned documentation are Markdown files that are added at the nested docs directy `/site/docs/docs/<version>/docs/*.md` files. Notice the location of the `mkdocs.yml`. Looking at this though, you may ask where the older versions and javadocs are.
```
Comment (RussellSpitzer) [inline site/README.md:41]: Not sure I understand the last two sentences of this paragraph, are they needed?

File: site/dev/common.sh [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+  assert_not_empty "${BRANCH}"
+
+  git push "${REMOTE}" "${BRANCH}" 
+}
+
+install_deps () {
+  pip -q install -r requirements.txt --upgrade
+}
+
+assert_not_empty () {
```
Comment (RussellSpitzer) [inline site/dev/common.sh:49]: not a huge deal but may help to pass through an error message so we know which arg is missing

Echo "Missing an argument: $2 "

**Review by RussellSpitzer — APPROVED** [review]
Saw a few nits in the docs, but worked well locally for me. Worktree code seems to working as expected but i'm not an expert in their usage for something like this.

File: site/mkdocs.yml (hunk 15 above)
Comment (RussellSpitzer) [reply site/mkdocs.yml:53]: For me I was wondering why "nightly" was first isn't that our "unstable" basically? 

I agree ideally we have something like 1.4.2 (latest)

Comment (bitsondatadev) [reply site/mkdocs.yml:53]: Yeah, so nightly is unstable. I thought it would be best to have it up top for the easiest access. Perhaps this ordering?:

```
1.4.2 (latest)
nightly
1.4.1
1.4.0
...
```

File: .github/workflows/site-ci.yml [hunk 20]
```diff
@@ -1,3 +1,4 @@
+#
```
Comment (Fokko) [inline .github/workflows/site-ci.yml:1]: Looks like git sees this as a move 🤔 Not a real problem.

File: .github/workflows/site-ci.yml [hunk 21]
```diff
@@ -14,13 +15,24 @@
...
+  deploy:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v4
+        with:
+          python-version: 3.x
+      - name: Deploy Iceberg documentation
+        run: |
+          make deploy
```
Comment (Fokko) [inline .github/workflows/site-ci.yml:37]: Nit:
```suggestion
        run: make deploy
```

File: site/dev/common.sh [hunk 22]
```diff
@@ -0,0 +1,228 @@
...
+  echo " --> pull versioned docs"
+  
+  # Ensure the remote repository for documentation exists and is up-to-date
+  create_or_update_docs_remote  
+
+  rm -r docs/docs
+
+  # Add local worktrees for documentation and javadoc from the remote repository
+  git worktree add docs/docs "${REMOTE}/docs"
+  git worktree add docs/javadoc "${REMOTE}/javadoc"
```
Comment (Fokko) [inline site/dev/common.sh:196]: I had still some old cruft on my worktree, so I needed to do this:
```suggestion
  git worktree add -f docs/docs "${REMOTE}/docs"
  git worktree add -f docs/javadoc "${REMOTE}/javadoc"
```

File: site/docs/community.md [hunk 23]
```diff
@@ -40,13 +40,13 @@ Issues are tracked in GitHub:
 
 ## Slack
 
-We use the [Apache Iceberg workspace](https://apache-iceberg.slack.com/) on Slack. To be invited, follow [this invite link](https://join.slack.com/t/apache-iceberg/shared_invite/zt-1znkcg5zm-7_FE~pcox347XwZE3GNfPg).
+We use the [Apache Iceberg workspace](https://apache-iceberg.slack.com/) on Slack. To be invited, follow [this invite link](https://join.slack.com/t/apache-iceberg/shared_invite/zt-287g3akar-K9Oe_En5j1UL7Y_Ikpai3A).
```
Comment (Fokko) [inline site/docs/community.md:43]: I checked and this is the latest one: https://github.com/apache/iceberg-docs/commit/87149e3da04eafad1c8766132f0bf1e822e4d075

File: site/docs/releases.md [hunk 24]
```diff
@@ -64,7 +67,140 @@ To add a dependency on Iceberg in Maven, add the following to your `pom.xml`:
 </dependencies>
 ```
 
-## 1.3.1 release
+### 1.4.3 Release
+
+Apache Iceberg 1.4.3 was released on December 27, 2023. The main issue it solves is missing files from a transaction retry with conflicting manifests. It is recommended to upgrade if you use transactions.
+
+- Core: Scan only live entries in partitions table (#8969) by @Fokko in [#9197](https://github.com/apache/iceberg/pull/9197)
+- Core: Fix missing files from transaction retries with conflicting manifest merges by [@nastra](https://github.com/nastra) in [#9337]O(https://github.com/apache/iceberg/pull/9337)
```
Comment (Fokko) [inline site/docs/releases.md:75]: This was caught by @manuzhang in https://github.com/apache/iceberg-docs/pull/299 🙌 

```suggestion
- Core: Fix missing files from transaction retries with conflicting manifest merges by [@nastra](https://github.com/nastra) in [#9337](https://github.com/apache/iceberg/pull/9337)
```

File: site/docs/spec.md [hunk 25]
```diff
@@ -1128,6 +1125,41 @@ Example
      ] } ]
 ```
 
+### Content File (Data and Delete) Serialization
```
Comment (Fokko) [inline site/docs/spec.md:1128]: Not for this PR, but it would be good at some point to just include the spec from the repository itself.

File: site/nav.yml [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+
+nav:
+  - Home: index.md
+  - Quickstart:
+    - Spark: spark-quickstart.md
+    - Hive: hive-quickstart.md
+  - Docs: 
+    #- nightly: '!include docs/docs/nightly/mkdocs.yml'
+    - latest: '!include docs/docs/latest/mkdocs.yml'
+    - 1.4.2: '!include docs/docs/1.4.2/mkdocs.yml'
```
Comment (Fokko) [inline site/nav.yml:26]: I think 1.4.3 is missing here.

File: site/docs/roadmap.md [hunk 27]
```diff
@@ -20,28 +20,37 @@ title: "Roadmap"
...
+* [Views Support](https://github.com/apache/iceberg/projects/29)
+* [Change Data Capture (CDC) Support](https://github.com/apache/iceberg/projects/26)
+* [Snapshot tagging and branching](https://github.com/apache/iceberg/projects/4)
+* [Inline file compaction](https://github.com/apache/iceberg/projects/14)
+* [Delete File compaction](https://github.com/apache/iceberg/projects/10)
+* [Z-ordering / Space-filling curves](https://github.com/apache/iceberg/projects/16)
+* [Support UPSERT](https://github.com/apache/iceberg/projects/15)
+
+# Clients
+_Rust and Go projects are pointing to their respective repositories which include
```
Comment (Fokko) [inline site/docs/roadmap.md:39]:
```suggestion
_Python, Rust and Go projects are pointing to their respective repositories which include 
```

**Review by Fokko — COMMENTED** [review]
Some small comments. I did a local build and it all works. The UI needs a bit of rework, but we can do that in a separate PR. What I do think is that we need to add 1.4.3 in https://github.com/apache/iceberg/tree/docs before merging this.

File: site/docs/community.md (hunk 23 above)
Comment (bitsondatadev) [reply site/docs/community.md:43]: Oh yeah, I still need to figure out how to embed jitsi stuff there, another day.

File: site/nav.yml (hunk 26 above)
Comment (bitsondatadev) [reply site/nav.yml:25]: I will update 1.4.3 on another PR

File: site/docs/community.md (hunk 23 above)
Comment (bitsondatadev) [reply site/docs/community.md:43]: wait yeah this is the latest one haha and that's what's currently in there.

**Comment by bitsondatadev** [comment]
@Fokko I updated after your comments!

**Comment by Fokko** [comment]
This looks great, and I don't see any open comments. If there is anything we can follow up on in a separate PR. I'll go ahead and rebase and merge this 🚀  Thanks for the reviews @rdblue, @RussellSpitzer, and a special thanks to @bitsondatadev for taking care of the docs!
